Test failures on 32-bit ARM build of FLINT 3.5.0

**Describe the bug**
On the 32-bit ARM build of the FLINT 3.5.0 package in Debian experimental, the tests failed with the following:

```
mpoly_test_irreducible...
FAIL: check 8 variable example
make[2]: *** [Makefile:792: build/mpoly/test/main_TEST_RUN] Aborted
make[2]: *** Waiting for unfinished jobs....
nmod_mpoly_add_sub...
nmod_mpoly_add_sub                                0.48   ([1;32mPASS[0m)
nmod_mpoly_add_sub_ui...
nmod_mpoly_add_sub_ui                             0.11   ([1;32mPASS[0m)
nmod_mpoly_cmp...
nmod_mpoly_cmp                                    0.08   ([1;32mPASS[0m)
nmod_mpoly_compose_nmod_mpoly...
nmod_mpoly_compose_nmod_mpoly                     0.00   ([1;31mFAIL[0m)
Check non-example 1
make[2]: *** [Makefile:792: build/nmod_mpoly/test/main_TEST_RUN] Aborted
fmpz_mpoly_add_sub...
fmpz_mpoly_add_sub                                0.64   ([1;32mPASS[0m)
fmpz_mpoly_add_sub_fmpz...
fmpz_mpoly_add_sub_fmpz                           0.11   ([1;32mPASS[0m)
fmpz_mpoly_add_sub_si...
fmpz_mpoly_add_sub_si                             0.11   ([1;32mPASS[0m)
fmpz_mpoly_add_sub_ui...
fmpz_mpoly_add_sub_ui                             0.11   ([1;32mPASS[0m)
fmpz_mpoly_buchberger_naive...
fmpz_mpoly_buchberger_naive                       0.03   ([1;32mPASS[0m)
fmpz_mpoly_cmp...
fmpz_mpoly_cmp                                    0.08   ([1;32mPASS[0m)
fmpz_mpoly_compose_fmpz_mpoly...
fmpz_mpoly_compose_fmpz_mpoly                     0.00   ([1;31mFAIL[0m)
Check non-example 1
make[2]: *** [Makefile:792: build/fmpz_mpoly/test/main_TEST_RUN] Aborted
```
Full log: https://buildd.debian.org/status/fetch.php?pkg=flint&arch=armhf&ver=3.5.0-1~exp1&stamp=1777239933&file=log

**Steps to reproduce**
This is from running `make -j8 check TESTSUITEFLAGS="-j8 --verbose" VERBOSE=1 V=1` after building FLINT 3.5.0 on a 32-bit ARM system running Debian experimental.

**Expected behavior**
The tests should pass

**System (please complete the following information):**
<ul>
<li>System CPU: armhf host on a an arm64 machine (arm-conova-01.debian.org)</li>
<li>Version of FLINT (if using Git, please specify commit): 3.5.0 (Debian package built from `v3.5.0` tag)</li>
<li>How FLINT was configured (i.e. options pushed): <code>./configure --build=arm-linux-gnueabihf --prefix=/usr --includedir=\${prefix}/include --mandir=\${prefix}/share/man --infodir=\${prefix}/share/info --sysconfdir=/etc --localstatedir=/var --disable-option-checking --disable-silent-rules --libdir=\${prefix}/lib/arm-linux-gnueabihf --runstatedir=/run --disable-maintainer-mode --disable-dependency-tracking
</code></li>
<li><details>
<summary>Output of <code>configure</code></summary>

```
checking build system type... arm-unknown-linux-gnueabihf
checking host system type... arm-unknown-linux-gnueabihf
checking how to print strings... printf
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc accepts -g... yes
checking for gcc option to enable C11 features... none needed
checking for a sed that does not truncate output... /usr/bin/sed
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /usr/bin/arm-linux-gnueabihf-ld
checking if the linker (/usr/bin/arm-linux-gnueabihf-ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking how to convert arm-unknown-linux-gnueabihf file names to arm-unknown-linux-gnueabihf format... func_convert_file_noop
checking how to convert arm-unknown-linux-gnueabihf file names to toolchain format... func_convert_file_noop
checking for /usr/bin/arm-linux-gnueabihf-ld option to reload object files... -r
checking for file... file
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ranlib... ranlib
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for gawk... no
checking for mawk... mawk
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for a working dd... /usr/bin/dd
checking how to truncate binary pipes... /usr/bin/dd bs=4096 count=1
checking for mt... no
checking if : is a manifest tool... no
checking for stdio.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for strings.h... yes
checking for sys/stat.h... yes
checking for sys/types.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... yes
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/arm-linux-gnueabihf-ld) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... no
checking for a race-free mkdir -p... /usr/bin/mkdir -p
checking how to run the C preprocessor... gcc -E
checking if compiler is GCC... yes
checking if compiler is Clang... no
checking for inline... inline
checking whether byte ordering is bigendian... no
checking if memory is strongly-ordered... unsure
checking for gmp.h... yes
checking if version of GMP is greater than 6.2.1... yes
checking if GMP defines mp_limb_t as unsigned long long int... no
checking for mpfr.h... yes
checking if version of MPFR is greater than 4.1.0... yes
checking for desired ABI... 32
checking for stdarg.h... yes
checking for math.h... yes
checking for float.h... yes
checking for errno.h... yes
checking for fenv.h... yes
checking for alloca.h... yes
checking for malloc.h... yes
checking for sys/param.h... yes
checking for windows.h... no
checking for pthread_np.h... no
checking for pthread.h... yes
checking for library containing atan2... -lm
checking for library containing __gmpz_init... -lgmp
checking for library containing __gmpn_mul_basecase... none required
checking for library containing __gmpn_gcd_11... none required
checking for library containing __gmpn_div_q... none required
checking for library containing __gmpn_invert_limb... none required
checking for library containing __gmpn_modexact_1_odd... no
checking for library containing __gmpn_addmul_2... no
checking for library containing __gmpn_add_n_sub_n... none required
checking for library containing __gmpn_add_nc... none required
checking for library containing __gmpn_sub_nc... none required
checking for library containing __gmpn_addlsh1_n... none required
checking for library containing __gmpn_addlsh1_n_ip1... no
checking for library containing __gmpn_rsh1sub_n... none required
checking for library containing __gmpn_rsh1add_n... none required
checking for library containing mpfr_init... -lmpfr
checking for library containing mpfr_round_p... none required
checking for library containing mpfr_mulhigh_n... none required
checking for library containing mpfr_sqrhigh_n... none required
checking whether gcc accepts -pthread... yes
checking if cpu_set_t is supported... yes
checking if alloca works... yes
checking for aligned_alloc... yes
checking for _aligned_malloc... no
checking whether gcc accepts -g -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/build/reproducible-path/flint-3.5.0=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security... yes
checking if system can use FLINT's fft_small module... no
configure: creating ./config.status
config.status: creating Makefile
config.status: creating flint.pc
config.status: creating src/flint.h
config.status: creating src/config.h
config.status: creating src/flint-config.h
config.status: linking src/gmpcompat.h.in to src/gmpcompat.h
config.status: linking src/mpn_extras/generic/flint-mparam.h to src/flint-mparam.h
config.status: linking src/fmpz/link/fmpz_single.c to src/fmpz/fmpz.c
config.status: executing libtool commands

```

</details></li></ul>

Getting one of the same failures on i386 (32-bit x86):

```
mpoly_test_irreducible...
FAIL: check 8 variable example
make[2]: *** [Makefile:792: build/mpoly/test/main_TEST_RUN] Aborted
make[2]: *** Waiting for unfinished jobs....
nmod_mpoly_add_sub...
nmod_mpoly_add_sub                                1.83   ([1;32mPASS[0m)
nmod_mpoly_add_sub_ui...
nmod_mpoly_add_sub_ui                             0.41   ([1;32mPASS[0m)
nmod_mpoly_cmp...
nmod_mpoly_cmp                                    0.37   ([1;32mPASS[0m)
nmod_mpoly_compose_nmod_mpoly...
nmod_mpoly_compose_nmod_mpoly                     0.00   ([1;31mFAIL[0m)
Check non-example 1
make[2]: *** [Makefile:792: build/nmod_mpoly/test/main_TEST_RUN] Aborted
```

Full log: https://buildd.debian.org/status/fetch.php?pkg=flint&arch=i386&ver=3.5.0-1~exp1&stamp=1777241080&file=log

Different test, but possibly related, on ppc64el:

```
make[2]: *** [Makefile:792: build/nmod_mpoly_factor/test/main_TEST_RUN] Segmentation fault
make[2]: *** Waiting for unfinished jobs....
fq_nmod_poly_factor                               0.75   ([1;32mPASS[0m)
```

Full log: https://buildd.debian.org/status/fetch.php?pkg=flint&arch=ppc64el&ver=3.5.0-1~exp1&stamp=1777241349&file=log

--- a/doc/source/history.rst
+++ b/doc/source/history.rst
@@ -7,6 +7,17 @@ FLINT version history
 :::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
 
 
+FLINT 3.6.0 (in development)
+-------------------------------------------------------------------------------
+
+Bug fixes
+
+* Fix ``flint_sprintf`` producing truncated output (e.g. ``"x"`` instead of
+  ``"x1"``) on 32-bit glibc, which broke ``mpoly_test_irreducible`` and the
+  ``compose_mpoly`` tests on i386/armhf
+  [EC, `#2648 <https://github.com/flintlib/flint/pull/2648>`_].
+
+
 2026-04-24 -- FLINT 3.5.0
 -------------------------------------------------------------------------------
 

--- a/src/generic_files/io_vsnprintf.c
+++ b/src/generic_files/io_vsnprintf.c
@@ -163,20 +163,3 @@ int flint_snprintf(char * s, size_t n, const char * str, ...)
 
    return ret;
 }
-
-int flint_vsprintf(char * s, const char * str, va_list vlist)
-{
-    return flint_vsnprintf(s, INT_MAX, str, vlist);
-}
-
-int flint_sprintf(char * s, const char * str, ...)
-{
-   va_list vlist;
-   int ret;
-
-   va_start(vlist, str);
-   ret = flint_vsprintf(s, str, vlist);
-   va_end(vlist);
-
-   return ret;
-}

--- a/src/generic_files/io_vsprintf.c
+++ b/src/generic_files/io_vsprintf.c
@@ -1,0 +1,136 @@
+/*
+    Copyright (C) 2026 Lars Göttgens
+    Copyright (C) 2026 Edgar Costa
+
+    This file is part of FLINT.
+
+    FLINT is free software: you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License (LGPL) as published
+    by the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.  See <https://www.gnu.org/licenses/>.
+*/
+
+#include <ctype.h> /* isdigit */
+#include <stdint.h> /* intmax_t */
+#include <stdio.h>
+#include <string.h> /* memcpy, strncmp and strchr */
+#include <stdarg.h>
+#include <wchar.h> /* wchar_t and wint_t */
+#include "nmod_types.h"
+#include "fmpz.h"
+#include "fmpz_mod.h"
+#include "fmpq_types.h"
+#include "fmpq.h"
+#include "arf_types.h"
+#include "arb.h"
+#include "acb.h"
+#include "gr.h"
+#include "gr_vec.h"
+#include "gr_poly.h"
+#include "gr_ore_poly.h"
+#include "gr_mat.h"
+#include "gr_mat/impl.h"
+
+/*
+    Sink for flint_sprintf-style functions: writes into a caller-supplied
+    buffer with no length bound. Uses system vsprintf rather than vsnprintf
+    because some 32-bit glibc versions corrupt output when vsnprintf is
+    called with a very large size_t (issue #2646).
+*/
+typedef struct
+{
+    char * buf;
+    size_t used;
+} flint_vsprintf_out;
+
+static void flint_vsprintf_init(flint_vsprintf_out * out, char * buf)
+{
+    out->buf = buf;
+    out->used = 0;
+    out->buf[0] = '\0';
+}
+
+static int flint_vsprintf_vprintf(flint_vsprintf_out * out, const char * fmt, va_list ap)
+{
+    va_list ap_copy;
+    int res;
+
+    va_copy(ap_copy, ap);
+    res = vsprintf(out->buf + out->used, fmt, ap_copy);
+    va_end(ap_copy);
+
+    if (res > 0)
+        out->used += (size_t) res;
+
+    return res;
+}
+
+static int flint_vsprintf_printf(flint_vsprintf_out * out, const char * fmt, ...)
+{
+    va_list ap;
+    int res;
+
+    va_start(ap, fmt);
+    res = flint_vsprintf_vprintf(out, fmt, ap);
+    va_end(ap);
+
+    return res;
+}
+
+static size_t flint_vsprintf_write(const void * buf, size_t len, flint_vsprintf_out * out)
+{
+    memcpy(out->buf + out->used, buf, len);
+    out->used += len;
+    out->buf[out->used] = '\0';
+    return len;
+}
+
+static int flint_vsprintf_putc(int ch, flint_vsprintf_out * out)
+{
+    out->buf[out->used] = (char) ch;
+    out->used++;
+    out->buf[out->used] = '\0';
+    return (unsigned char) ch;
+}
+
+#define FLINT_VPRINTF_FUNCTION flint_vsprintf
+#define FLINT_VPRINTF_FUNCTION_ARGS char * s
+#define FLINT_VPRINTF_OUT_T flint_vsprintf_out
+#define FLINT_VPRINTF_INIT(out) flint_vsprintf_init((out), s)
+#define FLINT_VPRINTF_PRINTF(out, ...) flint_vsprintf_printf((out), __VA_ARGS__)
+#define FLINT_VPRINTF_VPRINTF(out, fmt, ap) flint_vsprintf_vprintf((out), (fmt), (ap))
+#define FLINT_VPRINTF_WRITE(buf, len, out) flint_vsprintf_write((buf), (len), (out))
+#define FLINT_VPRINTF_PUTC(ch, out) flint_vsprintf_putc((ch), (out))
+#define FLINT_VPRINTF_PUTC_ERRVAL (-1)
+#define FLINT_VPRINTF_GR_STREAM_INIT(gr_out, out) gr_stream_init_str((gr_out))
+#define FLINT_VPRINTF_GR_STREAM_FLUSH(gr_out, out) \
+    do { \
+        FLINT_VPRINTF_WRITE((gr_out)->s, (gr_out)->len, (out)); \
+        flint_free((gr_out)->s); \
+    } while (0)
+
+#include "io_vprintf_impl.h"
+
+#undef FLINT_VPRINTF_FUNCTION
+#undef FLINT_VPRINTF_FUNCTION_ARGS
+#undef FLINT_VPRINTF_OUT_T
+#undef FLINT_VPRINTF_INIT
+#undef FLINT_VPRINTF_PRINTF
+#undef FLINT_VPRINTF_VPRINTF
+#undef FLINT_VPRINTF_WRITE
+#undef FLINT_VPRINTF_PUTC
+#undef FLINT_VPRINTF_PUTC_ERRVAL
+#undef FLINT_VPRINTF_GR_STREAM_INIT
+#undef FLINT_VPRINTF_GR_STREAM_FLUSH
+
+int flint_sprintf(char * s, const char * str, ...)
+{
+   va_list vlist;
+   int ret;
+
+   va_start(vlist, str);
+   ret = flint_vsprintf(s, str, vlist);
+   va_end(vlist);
+
+   return ret;
+}

--- a/src/test/main.c
+++ b/src/test/main.c
@@ -36,6 +36,7 @@ test_struct tests[] =
     TEST_FUNCTION(flint_ctz),
     TEST_FUNCTION(flint_fprintf),
     TEST_FUNCTION(flint_snprintf),
+    TEST_FUNCTION(flint_sprintf),
     TEST_FUNCTION(flint_printf),
     TEST_FUNCTION(memory_manager),
     TEST_FUNCTION(sdiv_qrnnd),

--- a/src/test/t-io.c
+++ b/src/test/t-io.c
@@ -1450,7 +1450,7 @@ TEST_FUNCTION_START(flint_sprintf, state)
         char expected[16];
         memset(buf, 'Z', sizeof(buf));
         got = flint_sprintf(buf, "x%wd", i + 1);
-        snprintf(expected, sizeof(expected), "x%ld", (long) (i + 1));
+        snprintf(expected, sizeof(expected), "x%lld", (long long) (i + 1));
         if (got < 0)
             TEST_FUNCTION_FAIL(
                     "Negative return value from flint_sprintf for i=%wd.\n", i);
@@ -1468,7 +1468,7 @@ TEST_FUNCTION_START(flint_sprintf, state)
             char expected[64];
             memset(buf, 'Z', sizeof(buf));
             got = flint_sprintf(buf, "[%wd]", values[ix]);
-            snprintf(expected, sizeof(expected), "[%ld]", (long) values[ix]);
+            snprintf(expected, sizeof(expected), "[%lld]", (long long) values[ix]);
             if (got < 0 || strcmp(buf, expected) != 0)
                 TEST_FUNCTION_FAIL(
                         "flint_sprintf(\"[%%wd]\", %wd) gave \"%s\" expected \"%s\"\n",

--- a/src/test/t-io.c
+++ b/src/test/t-io.c
@@ -1435,6 +1435,58 @@ TEST_FUNCTION_START(flint_snprintf, state)
     TEST_FUNCTION_END(state);
 }
 
+TEST_FUNCTION_START(flint_sprintf, state)
+{
+    /* Regression test for #2646: on 32-bit glibc, vsnprintf called with a
+       very large size_t (such as INT_MAX) silently drops output past the
+       first character, causing flint_sprintf("x%wd", n) to produce "x"
+       instead of "x<n>". */
+    char buf[64];
+    int got;
+    slong i;
+
+    for (i = 0; i < 10; i++)
+    {
+        char expected[16];
+        memset(buf, 'Z', sizeof(buf));
+        got = flint_sprintf(buf, "x%wd", i + 1);
+        snprintf(expected, sizeof(expected), "x%ld", (long) (i + 1));
+        if (got < 0)
+            TEST_FUNCTION_FAIL(
+                    "Negative return value from flint_sprintf for i=%wd.\n", i);
+        if (strcmp(buf, expected) != 0)
+            TEST_FUNCTION_FAIL(
+                    "flint_sprintf(\"x%%wd\", %wd) gave \"%s\" expected \"%s\"\n",
+                    i + 1, buf, expected);
+    }
+
+    {
+        slong values[] = {0, 1, -1, 12345, -12345, WORD_MIN, WORD_MAX};
+        size_t ix;
+        for (ix = 0; ix < sizeof(values) / sizeof(values[0]); ix++)
+        {
+            char expected[64];
+            memset(buf, 'Z', sizeof(buf));
+            got = flint_sprintf(buf, "[%wd]", values[ix]);
+            snprintf(expected, sizeof(expected), "[%ld]", (long) values[ix]);
+            if (got < 0 || strcmp(buf, expected) != 0)
+                TEST_FUNCTION_FAIL(
+                        "flint_sprintf(\"[%%wd]\", %wd) gave \"%s\" expected \"%s\"\n",
+                        values[ix], buf, expected);
+        }
+    }
+
+    {
+        memset(buf, 'Z', sizeof(buf));
+        got = flint_sprintf(buf, "a%wd b%wu c%wx", (slong) -7, (ulong) 42, (ulong) 0xab);
+        if (got < 0 || strcmp(buf, "a-7 b42 cab") != 0)
+            TEST_FUNCTION_FAIL(
+                    "flint_sprintf with multiple %%w specifiers: got \"%s\"\n", buf);
+    }
+
+    TEST_FUNCTION_END(state);
+}
+
 
 #if HAVE_UNISTD_H && FLINT_COVERAGE
 #include <unistd.h>

